io-uring/src/ring.rs

```diff
@@ -0,0 +1,240 @@
+    pub fn process_completions(&mut self) -> io::Result<()> {
+        let mut completion = self.ring.completion();
+        let mut new_entries = smallvec![];
+        loop {
+            let Some(cqe) = completion.next() else {
+                break;
+            };
+            let completed_key = cqe.user_data() as usize;
+            let entry = self.entries.get_mut(completed_key).unwrap();
+            let result = entry.result(cqe.result());
+            let mut comp_ctx = Completion {
+                context: &mut self.context,
+                new_entries,
+            };
+            let res = entry.complete(&mut comp_ctx, result);
+            if !cqueue::more(cqe.flags()) {
+                self.entries.remove(completed_key);
+            }
+            res?;
+            new_entries = std::mem::take(&mut comp_ctx.new_entries);
+            if !new_entries.is_empty() {
+                completion.sync();
+                drop(completion);
+                for new_entry in new_entries.drain(..) {
+                    self.push(new_entry)?;
+                }
+                completion = self.ring.completion();
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Drains the ring.
+    ///
+    /// This will submit all pending operations to the kernel and process all
+    /// completions until the ring is empty.
+    pub fn drain(&mut self) -> io::Result<()> {
+        loop {
+            self.process_completions()?;
+
+            if self.entries.is_empty() {
+                break;
+            }
+
+            match self.ring.submitter().submit_with_args(
+                1,
+                &SubmitArgs::new().timespec(&Timespec::from(Duration::from_millis(10))),
+            ) {
+                Ok(_) => {}
+                Err(e) if e.raw_os_error() == Some(libc::ETIME) => {}
+                Err(e) => return Err(e),
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Trait for operations that can be submitted to a [Ring].
+pub trait RingOp<T> {
+    fn entry(&mut self) -> squeue::Entry;
+    fn complete(&mut self, ctx: &mut Completion<T, Self>, res: io::Result<i32>) -> io::Result<()>
+    where
+        Self: Sized;
+    fn result(&self, res: i32) -> io::Result<i32> {
+        if res < 0 {
+            Err(io::Error::from_raw_os_error(res.wrapping_neg()))
+        } else {
+            Ok(res)
+        }
+    }
+}
+
+/// Context object passed to [RingOp::complete].
+pub struct Completion<'a, T, E: RingOp<T>> {
+    // Give new_entries a stack size of 2 to avoid heap allocations in the common case where only
+    // one or two ops are queued from a completion handler.
+    //
+    // It's common to want to queue some extra work after a completion, for instance if you've
+    // completed a read and want to close the file descriptor, or if you're doing chained operations
+    // and want to push the next one. It's less common to want to queue many operations.
+    new_entries: SmallVec<[E; 2]>,
```

Comment:
Can we make 2 a generic const param?

Reply:
Ok. and it turns out it is not easy to push the const 2 to generics, because SmallVec requires Array trait for inner type and Array trait is not implemented for generic const array.